src/crypto/equihash.cpp

```diff
@@ -514,8 +639,12 @@ bool Equihash<N,K>::OptimisedSolve(const eh_HashState& base_state,
             icv.reserve(recreate_size);
             for (eh_index j = 0; j < recreate_size; j++) {
                 eh_index newIndex { UntruncateIndex(partialSoln.get()[i], j, CollisionBitLength + 1) };
-                icv.emplace_back(base_state, newIndex, HashLength,
-                                 CollisionBitLength, CollisionByteLength);
+                if (j == 0 || newIndex % IndicesPerHashOutput == 0) {
```

Comment:
The indices are big-endian (for the purposes of being in rows), and we truncate the least significant bits (so that sorting the truncated indices is approximately the same as sorting the full indices). Thus we are guaranteed to get contiguous indices inside that loop, and can generate a hash once per `IndicesPerHashOutput` indices. The `j == 0` is to ensure we have a hash to start with if the first index doesn't fall on a hash boundary (which is not guaranteed).